tensorflow/c/experimental/gradients/math_grad.cc

```diff
@@ -57,21 +69,142 @@ class ExpGradientFunction : public GradientFunction {
   Status Compute(Context* ctx,
                  absl::Span<AbstractTensorHandle* const> grad_inputs,
                  vector<AbstractTensorHandle*>* grad_outputs) override {
-    vector<AbstractTensorHandle*> conj_outputs(1);
-    TF_RETURN_IF_ERROR(
-        Conj(ctx->ctx, {exp_.get()}, absl::MakeSpan(conj_outputs), "ExpConj"));
+    std::vector<AbstractTensorHandle*> conj_outputs(1);
+    std::string name = "Conj_Exp_Grad_" + std::to_string(counter);
+    TF_RETURN_IF_ERROR(Conj(ctx->ctx, {exp_.get()},
+                            absl::MakeSpan(conj_outputs), name.c_str()));
     AbstractTensorHandlePtr conj_output_releaser(conj_outputs[0]);
     grad_outputs->resize(1);
+
+    name = "Mul_Exp_Grad_" + std::to_string(counter);
     TF_RETURN_IF_ERROR(Mul(ctx->ctx, {conj_outputs[0], grad_inputs[0]},
-                           absl::MakeSpan(*grad_outputs), "ExpGradMul"));
+                           absl::MakeSpan(*grad_outputs), name.c_str()));
     return Status::OK();
   }
   ~ExpGradientFunction() override {}
 
  private:
+  int64_t counter;
   AbstractTensorHandlePtr exp_;
 };
 
+class MatMulGradientFunction : public GradientFunction {
+ public:
+  explicit MatMulGradientFunction(std::vector<AbstractTensorHandle*> f_inputs,
+                                  AttrBuilder f_attrs)
+      : forward_inputs(f_inputs), attrs(f_attrs) {}
+
+  Status Compute(Context* ctx,
+                 absl::Span<AbstractTensorHandle* const> grad_inputs,
+                 std::vector<AbstractTensorHandle*>* grad_outputs) override {
+    /* Given upstream grad U and a matmul op A*B, the gradients are:
+     *
+     *    dA = U * B.T
+     *    dB = A.T * U
+     *
+     *    where A.T means `transpose(A)`
+     */
+
+    // TODO(amturati): figure why adding attrs to the function breaks the
```

Comment:
Could you explain a bit more about the problem you are running into here?

Reply:
Adding the `attrs` for some reason makes the counter fail to update sometimes or take on very strange large values. I tried making `counter` a static variable so every reference to it updates the same way, but this gives me a compile error: "may overflow at runtime; recompile with -fPIC"

Reply:
Which `attrs` are you referring to?

Reply:
`attrs` are passed in to the MatMul gradient constructor to get the `transpose` atrributes from the forward pass